app/src/main/java/org/stepic/droid/notifications/StepicFcmListenerService.kt

```diff
@@ -5,39 +5,61 @@ import com.google.firebase.messaging.RemoteMessage
 import org.stepic.droid.base.App
+import org.stepic.droid.notifications.badges.NotificationsBadgesManager
 import org.stepic.droid.notifications.model.Notification
+import org.stepic.droid.notifications.model.NotificationStatuses
 import org.stepic.droid.preferences.SharedPreferenceHelper
 import java.io.IOException
 import javax.inject.Inject
 
 class StepicFcmListenerService : FirebaseMessagingService() {
+    companion object {
+        private const val NOTIFICATION_TYPE = "notifications"
+        private const val NOTIFICATION_STATUSES_TYPE ="notification-statuses"
+    }
 
-    val hacker = HackFcmListener()
+    private val hacker = HackFcmListener()
 
     override fun onMessageReceived(message: RemoteMessage) {
-        val data = message.data
-        val notificationRawString: String? = data?.get("object")
         try {
-            val userId: Long? = hacker.sharedPreferenceHelper.profile?.id
-            val userIdServerString = data?.get("user_id") ?: ""
-            val userIdServer = Integer.parseInt(userIdServerString)
-            if (userId == null || userIdServer.toLong() != userId) {
-                return;
-            }
-            val stepicNotification = Gson().fromJson(notificationRawString, Notification::class.java)
-            stepicNotification?.let {
-                hacker.stepikNotificationManager.showNotification(it)
+            message.data?.let { data ->
+                val userId: Long? = hacker.sharedPreferenceHelper.profile?.id
+                val userIdServerString = data["user_id"] ?: ""
+                val userIdServer = Integer.parseInt(userIdServerString)
+                if (userId == null || userIdServer.toLong() != userId) {
+                    return
+                }
+
+                val rawMessageObject = data["object"]
+                when (data["type"]) {
+                    NOTIFICATION_TYPE -> processNotification(rawMessageObject)
+                    NOTIFICATION_STATUSES_TYPE -> processNotificationStatuses(rawMessageObject)
+                }
             }
         } catch (e: IOException) {
             //internet is not available. just ignore, we can't show reach notification
         } catch (e: Exception) {
-            hacker.analytic.reportError(Analytic.Error.NOTIFICATION_ERROR_PARSE, e);
+            hacker.analytic.reportError(Analytic.Error.NOTIFICATION_ERROR_PARSE, e)
+        }
+    }
+
+    private fun processNotification(rawMessageObject: String?) {
+        val stepikNotification = Gson().fromJson(rawMessageObject, Notification::class.java)
+        stepikNotification?.let {
+            hacker.stepikNotificationManager.showNotification(it)
+        }
+    }
+
+    private fun processNotificationStatuses(rawMessageObject: String?) {
+        val notificationStatuses = Gson().fromJson(rawMessageObject, NotificationStatuses::class.java)
+        notificationStatuses?.let {
+            hacker.notificationsBadgesManager.syncCounter()
```

Comment:
we know here number of badges from `notificationStatuses` object. Why we should go to api again in `syncCounter`? Let's reuse this value?

Reply:
We can't because those notifications come up with delays and sometimes don't come at all